hazelcast/src/main/java/com/hazelcast/query/SqlPredicate.java

```diff
@@ -317,7 +317,7 @@ private void readObject(java.io.ObjectInputStream in)
         Predicate[] predicates;
         if (klass.isInstance(predicateLeft)) {
             subpredicatesLeft = ((CompoundPredicate) predicateLeft).getPredicates();
-            if (predicateRight instanceof CompoundPredicate) {
+            if (predicateRight.getClass() == klass) {
```

Comment:
I intend to add a couple of tests before we get this merged, so I will mark as 'dont merge' this PR for now. The problem was that sub-predicates joined by different logical operator (AND / OR) would be eventually flattened to a single compound predicate, discarding one of the two operators, leading to incorrect representation of the SQL statement. The fixed condition ensures we only flatten when operator is the same.